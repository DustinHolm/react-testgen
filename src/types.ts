export interface ImportBlock {
    sourceFile: string,
    isInternal: boolean,
    imports: Import[]
}

export interface Import {
    isDefault: boolean,
    name: string,
    element?: ExportableElement
}

export interface Export {
    name: string,
    isDefault: boolean,
    element?: ExportableElement
}

export enum ExportType {
    Class,
    Function,
    Interface,
    Variable
}

export interface ExportableElement {
    type: ExportType,
    returnsJSX: boolean
}

export interface ClassElement extends ExportableElement {
    methods: MethodElement[],
    constructors: Constructor[]
}

export interface MethodElement extends FunctionElement {
    name: string
}

export interface FunctionElement extends ExportableElement {
    returnType: string,
    parameters: Attribute[]
}

export interface InterfaceElement extends ExportableElement {
    attributes: Attribute[]
}

export interface VariableElement extends ExportableElement {
    varType: string,
    varValue: string
}

export interface Attribute {
    name: string,
    type: string
}

export interface Constructor {
    parameters: Attribute[]
}