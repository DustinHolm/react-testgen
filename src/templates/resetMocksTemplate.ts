import { Attribute, ImportBlock } from "../types"
import { toPascalCase } from "./common"

export const createResetMocksFunction = (props: Attribute[], mocks: ImportBlock[]): string => {
    const propResetCalls = props.map(a =>
        "    resetMockProp" + toPascalCase(a.name) + "()"
    ).join("\n")

    const mockResetCalls = mocks.flatMap(ib => ib.imports).map(i =>
        "    resetMock" + toPascalCase(i.name) + "()"
    ).join("\n")

    const newLine = propResetCalls.length > 0 && mockResetCalls.length > 0
        ? "\n"
        : ""

    return `\
function resetMocks() {
${propResetCalls}${newLine}${mockResetCalls}
}
`
}