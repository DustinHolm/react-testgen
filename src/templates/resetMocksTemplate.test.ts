import { Attribute, Export, ImportBlock } from "../types"
import { createResetMocksFunction } from "./resetMocksTemplate"

describe("resetMocksTemplate Tests", () => {
    test("With only props", () => {
        const props: Attribute[] = [{name: "att_1", type: "type_1"}, {name: "att_2", type: "type_2"}]
        const mocks: ImportBlock[] = []
        const returnedString = createResetMocksFunction(props, mocks)
        expect(returnedString).toMatchSnapshot()
    })

    test("With one mocked import block", () => {
        const props: Attribute[] = []
        const mocks: ImportBlock[] = [getImportBlock(2)]
        const returnedString = createResetMocksFunction(props, mocks)
        expect(returnedString).toMatchSnapshot()
    })

    test("With two mocked import blocks", () => {
        const props: Attribute[] = []
        const mocks: ImportBlock[] = [getImportBlock(1), getImportBlock(1)]
        const returnedString = createResetMocksFunction(props, mocks)
        expect(returnedString).toMatchSnapshot()
    })

    test("With mixed inputs", () => {
        const props: Attribute[] = [{name: "att_1", type: "type_1"}, {name: "att_2", type: "type_2"}]
        const mocks: ImportBlock[] = [getImportBlock(2)]
        const returnedString = createResetMocksFunction(props, mocks)
        expect(returnedString).toMatchSnapshot()
    })
})

function getImportBlock(numberOfImports: number): ImportBlock {
    const imports = []
    for (let i = 0; i < numberOfImports; i++) {
        imports.push(getImport(i))
    }

    return {
        sourceFile: "./mock",
        isInternal: true,
        imports: imports
    }
}

function getImport(num: number): Export {
    return {
        name: "import_" + num,
        isDefault: false
    }
}