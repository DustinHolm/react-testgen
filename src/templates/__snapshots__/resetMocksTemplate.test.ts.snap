// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resetMocksTemplate Tests With mixed inputs 1`] = `
"function resetMocks() {
    resetMockPropAtt_1()
    resetMockPropAtt_2()
    resetMockImport_0()
    resetMockImport_1()
}
"
`;

exports[`resetMocksTemplate Tests With one mocked import block 1`] = `
"function resetMocks() {
    resetMockImport_0()
    resetMockImport_1()
}
"
`;

exports[`resetMocksTemplate Tests With only props 1`] = `
"function resetMocks() {
    resetMockPropAtt_1()
    resetMockPropAtt_2()
}
"
`;

exports[`resetMocksTemplate Tests With two mocked import blocks 1`] = `
"function resetMocks() {
    resetMockImport_0()
    resetMockImport_0()
}
"
`;
